modules/core/src/main/scala/org/tessellation/domain/snapshot/programs/Download.scala

```diff
@@ -49,28 +53,52 @@ object Download {
     globalSnapshotContextFns: GlobalSnapshotContextFunctions[F],
     nodeStorage: NodeStorage[F],
     consensus: SnapshotConsensus[F, GlobalIncrementalSnapshot, GlobalSnapshotContext, _],
-    peerSelect: PeerSelect[F]
+    peerSelect: PeerSelect[F],
+    config: AppConfig
   ): Download[F] = new Download[F] {
 
     val logger = Slf4jLogger.getLogger[F]
 
     val minBatchSizeToStartObserving: Long = 1L
     val observationOffset = NonNegLong(4L)
     val fetchSnapshotDelayBetweenTrials = 10.seconds
+    val checkJoinablePeersDelayBetweenAttempts = 1.minute
+    val retryCheckJoinablePeersMax = 10.minutes
+    val lowerSelectablePeersCountThreshold: PosInt = 2
+    val upperSelectablePeersCountThreshold: PosInt = 99
+    val fallbackSelectablePeersCountThreshold: PosInt = 1
 
     type DownloadResult = (Signed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)
     type ObservationLimit = SnapshotOrdinal
 
-    def download: F[Unit] =
-      nodeStorage
-        .tryModifyState(NodeState.WaitingForDownload, NodeState.DownloadInProgress, NodeState.WaitingForObserving)(start)
-        .flatMap(observe)
-        .flatMap { result =>
-          val ((snapshot, context), observationLimit) = result
+    def waitForPeers(minPeers: PosInt): F[Set[Peer]] = retryingOnFailures[Set[Peer]](
+      limitRetriesByCumulativeDelay(
+        threshold = retryCheckJoinablePeersMax,
+        policy = constantDelay(checkJoinablePeersDelayBetweenAttempts)
+      ),
+      a => (a.size >= minPeers).pure[F],
+      (a, details) => logger.info(s"Discovered ${a.size} selectable peers, waiting for more: $details")
+    )(clusterStorage.getResponsivePeers.map(_.filter(_.state === Ready)))
+
+    def delayDownload: F[Unit] =
+      if (config.environment === Dev)
+        download
+      else
+        waitForPeers(lowerSelectablePeersCountThreshold) >>
+          waitForPeers(upperSelectablePeersCountThreshold) >>
```

Comment:
so we wait for 2 peers, then we wait for 99 peers. shouldn't it be a single operation only (waiting to 99)?

Reply:
@ryle-goehausen asked for a delay after 2 peers had been discovered. `waitForPeers(upperSelectablePeersCountThreshold)` was meant to be that delay. It purpose is simply to delay the download furthur. It has been replaced by `Async[F].sleep(additionalDiscoveryDelay)`.

Reply:
No doubt about it, but running it sequentially it basically means you wait for 99 peers anyway and the previous operations is kinda ignored.

Reply:
The purpose of `waitForPeers(upperSelectablePeersCountThreshold)` was to simply delay for 10 minutes. It has been replaced by `Async[F].sleep(additionalDiscoveryDelay)`.

Reply:
but `waitForPeers` didn't simply delay for 10 minutes but also check for the number of peers and continued if reached

Reply:
It has been replaced by `Async[F].sleep(additionalDiscoveryDelay)`.